panel/io/compile.py

```diff
@@ -0,0 +1,177 @@
+        if key.endswith('/'):
+            key = key[:-1]
+
+        if match:
+            dependencies[key] = f"^{match.group(1)}"
+        else:
+            dependencies[key] = "latest"
+
+    esm = component._render_esm(compiled=False)
+    if issubclass(component, ReactComponent):
+        esm += '\nimport * as React from "react"\nimport { createRoot } from "react-dom/client"\nexport default {render, React, createRoot}'
+
+    js_code, packages = extract_packages(esm)
+    dependencies.update(packages)
+
+    # Create package.json structure
+    package_json = {
+        "name": camel_to_kebab(component.__name__),
+        "dependencies": dependencies
+    }
+    return js_code, package_json
+
+
+def compile_component(component: type[ReactiveESM], minify: bool = True):
+    """
+    Compiles and bundles the ESM code of a ReactiveESM component and
+    writes the compiled JS file to a file adjacent to the module
+    the component was defined in.
+    """
+    name = camel_to_kebab(component.__name__)
+    ext = 'jsx' if issubclass(component, ReactComponent) else 'js'
+    if hasattr(component, '__path__'):
+        out_path = component.__path__
+    else:
+        out_path = pathlib.Path(inspect.getfile(component)).parent
+    out = (pathlib.Path(out_path) / f'{name}.compiled.js').absolute()
+
+    code, package_json = extract_dependencies(component)
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        package_json_path = os.path.join(temp_dir, 'package.json')
+        with open(package_json_path, 'w') as package_json_file:
+            json.dump(package_json, package_json_file, indent=2)
+
+        index_js_path = os.path.join(temp_dir, f'index.{ext}')
+        with open(index_js_path, 'w') as index_js_file:
+            index_js_file.write(code)
+
+        os.chdir(temp_dir)
+
+        install_cmd = ['npm', 'install']
+        try:
+            print(f"Running command: {' '.join(install_cmd)}")  # noqa
+            result = subprocess.run(install_cmd, check=True, capture_output=True, text=True)
+            print("esbuild output:", result.stdout)  # noqa
+            if result.stderr:
+                print("esbuild errors:", result.stderr)  # noqa
+        except subprocess.CalledProcessError as e:
+            print("An error occurred while running esbuild:")  # noqa
+            print(e.stderr)  # noqa
+
+        minify_arg = ['--minify'] if minify else []
+        build_cmd = ['esbuild', index_js_path, '--bundle', '--format=esm', f'--outfile={out}'] + minify_arg
+        try:
+            print(f"Running command: {' '.join(build_cmd)}")  # noqa
+            result = subprocess.run(build_cmd, check=True, capture_output=True, text=True)
+            print("esbuild output:", result.stdout)  # noqa
+            if result.stderr:
+                print("esbuild errors:", result.stderr)  # noqa
+        except subprocess.CalledProcessError as e:
+            print("An error occurred while running esbuild:")  # noqa
```

Comment:
If you want to spice it up with colors, you can take a look at `hatch_build.py`

Reply:
Done